### PR TITLE
Fix incremental sync hive test with fast sync

### PR DESCRIPTION
**Pull request by asdacap**
Fix incremental sync when fast sync is enabled by adding fast sync lag post merge.

## Changes:
- Added `lastBlockToIgnore` param to chain level helper get headers.
  - Implemented by fetching extra header, then removing them before returning the header list.
- In multisyncmodeselector, Replace `best.Peer.Block` to `best.TargetBlock` which is the `BeaconPivot.ProcessDestination`. 

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [X] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Other (please describe): 

## Testing
**Requires testing**

- [X] Yes
- [ ] No

**In case you checked yes, did you write tests??**

- [X] Yes
- [ ] No

**Comments about testing , should you have some** (optional)

- Hive test passes. Sync test all passed. Only one from transition test as expected.
- Goerli can sync to head.

## Further comments (optional)

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: src/Nethermind/Nethermind.Synchronization/ParallelSync/MultiSyncModeSelector.cs [hunk 1]
```diff
@@ -293,13 +293,13 @@ private void UpdateSyncModes(SyncMode newModes, string? reason = null)
         /// <param name="best">Snapshot of the best known states</param>
         /// <returns>A string describing the state of sync</returns>
         private static string BuildStateString(Snapshot best) =>
-            $"processed:{best.Processed}|state:{best.State}|block:{best.Block}|header:{best.Header}|chain difficulty:{best.ChainDifficulty}|peer block:{best.Peer.Block}|peer total difficulty:{best.Peer.TotalDifficulty}";
+            $"processed:{best.Processed}|state:{best.State}|block:{best.Block}|header:{best.Header}|chain difficulty:{best.ChainDifficulty}|peer block:{best.TargetBlock}|peer block:{best.Peer.Block}|peer total difficulty:{best.Peer.TotalDifficulty}";
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Synchronization/ParallelSync/MultiSyncModeSelector.cs:296]: **peer block**:{best.TargetBlock}|peer block:{best.Peer.Block}

change to target block

File: src/Nethermind/Nethermind.Synchronization/ParallelSync/MultiSyncModeSelector.cs [hunk 2]
```diff
@@ -293,13 +293,13 @@ private void UpdateSyncModes(SyncMode newModes, string? reason = null)
...
 
         private bool IsInAStickyFullSyncMode(Snapshot best)
         {
             long bestBlock = Math.Max(best.Processed, LastBlockThatEnabledFullSync ?? 0);
             bool hasEverBeenInFullSync = bestBlock > _pivotNumber && best.State > _pivotNumber;
-            long heightDelta = best.Peer.Block - bestBlock;
+            long heightDelta = best.TargetBlock - bestBlock;
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Synchronization/ParallelSync/MultiSyncModeSelector.cs:302]: these changes seems to be okay, but on the other hand we should test it a lot before merging

File: src/Nethermind/Nethermind.Synchronization/ParallelSync/MultiSyncModeSelector.cs [hunk 3]
```diff
@@ -555,8 +562,8 @@ private bool ShouldBeInStateSyncMode(Snapshot best)
 
         private bool CanBeInSnapRangesPhase(Snapshot best)
         {
-            long peerBlock = best.Peer.Block;
-            bool isCloseToHead = peerBlock >= best.Header && (peerBlock - best.Header) < Constants.MaxDistanceFromHead;
+            bool isCloseToHead = best.TargetBlock >= best.Header &&
```
Comment (MarekM25) [inline src/Nethermind/Nethermind.Synchronization/ParallelSync/MultiSyncModeSelector.cs:565]: so before we were closeToTheHead when we were close to the Peer. However, now we will relay on lastNewPayload, so we can start snap sync before we reach the head of the chain. Requires lots of testing :)

**Review by MarekM25 — COMMENTED** [review]
How was it tested so far? Hive tests of course, but syncing?

File: src/Nethermind/Nethermind.Merge.Plugin/Synchronization/ChainLevelHelper.cs [hunk 4]
```diff
@@ -123,12 +124,22 @@ public class ChainLevelHelper : IChainLevelHelper
...
         }
 
+        int toTake = headers.Count - lastBlockToIgnore;
+        if (toTake <= 0)
+        {
+            headers.Clear();
+        }
+        else
+        {
+            headers.RemoveRange(toTake, headers.Count - toTake);
```
[the remark this answers is not in the document]
Comment (asdacap) [reply src/Nethermind/Nethermind.Merge.Plugin/Synchronization/ChainLevelHelper.cs:140]: Yea, we don't exactly know how many block is needed. ProcessingTarget does not include beacon body from NP.

**Comment by asdacap** [comment]
Goerli can sync to head.

**Comment by asdacap** [comment]
Rinkeby can sync to head.

**Comment by asdacap** [comment]
All 4 mainnet smoke can sync to processing.

**Review by MarekM25 — APPROVED** [review]

File: src/Nethermind/Nethermind.Merge.Plugin/Synchronization/ChainLevelHelper.cs [hunk 5]
```diff
@@ -28,7 +28,7 @@ namespace Nethermind.Merge.Plugin.Synchronization;
 
 public interface IChainLevelHelper
 {
-    BlockHeader[]? GetNextHeaders(int maxCount, long maxHeaderNumber);
```
Comment (LukaszRozmej) [inline src/Nethermind/Nethermind.Merge.Plugin/Synchronization/ChainLevelHelper.cs:31]: Small:
- Naming, can we have something easier to decipher? Dunno maybe `skipLastBlockCount`?
- Maybe should be 0 by default?

**Review by LukaszRozmej — APPROVED** [review]
Hard to comprehend without running it and just testing, but the code looks ok.

File: src/Nethermind/Nethermind.Merge.Plugin/Synchronization/ChainLevelHelper.cs (hunk 5 above)
Comment (asdacap) [reply src/Nethermind/Nethermind.Merge.Plugin/Synchronization/ChainLevelHelper.cs:31]: Yea, my bad. Intellij refactor.

File: src/Nethermind/Nethermind.Synchronization/ParallelSync/MultiSyncModeSelector.cs (hunk 3 above)
Comment (LukaszRozmej) [reply src/Nethermind/Nethermind.Synchronization/ParallelSync/MultiSyncModeSelector.cs]: Ah good catch, we potentially might have issues when our CL is behind?

**Pull request merged**